### PR TITLE
Java: Replace MethodAccess, LValue, RValue with more intuitive names. Introduce NewClassExpr.

### PR DESCRIPTION
**Pull request by smowton**
Recommend reviewing commit-by-commit, which separates the mass-renames from the substantial changes.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/Expr.qll [hunk 1]
```diff
@@ -1831,37 +1844,43 @@ class ExtensionReceiverAccess extends VarAccess {
...
+ * A write access to a variable, which occurs as the destination of an assignment.
  */
-class LValue extends VarAccess {
-  LValue() { this.isLValue() }
+class VarWrite extends VarAccess {
+  VarWrite() { this.isVarWrite() }
 
   /**
-   * Gets a source expression used in an assignment to this l-value.
+   * Gets the right-hand side of the assignment that executes this variable write.
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Expr.qll:1853]: This change to the top-level one-liner is incorrect, as is also explained a few lines below, i.e. compound assignments include the LHS as a source. It's quite unfortunate that the predicate name is `getRhs`, as that's similarly wrong.

File: java/ql/lib/semmle/code/java/Expr.qll [hunk 2]
```diff
@@ -1831,37 +1844,43 @@ class ExtensionReceiverAccess extends VarAccess {
...
-class RValue extends VarAccess {
-  RValue() { this.isRValue() }
+class VarRead extends VarAccess {
+  VarRead() { this.isVarRead() }
 }
 
+/** DEPRECATED: Alias for `VarRead`. */
+deprecated class RValue = VarRead;
+
 /** A method access is an invocation of a method with a list of arguments. */
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Expr.qll:1882]:
```suggestion
/** A method call is an invocation of a method with a list of arguments. */
```

File: java/ql/lib/semmle/code/java/Expr.qll [hunk 3]
```diff
@@ -1912,18 +1931,27 @@ class MethodAccess extends Expr, Call, @methodaccess {
    * Holds if this is a method access to an instance method of `this`. That is,
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Expr.qll:1931]:
```suggestion
   * Holds if this is a method call to an instance method of `this`. That is,
```

File: java/ql/lib/semmle/code/java/JDK.qll [hunk 4]
```diff
@@ -261,8 +261,8 @@ class MethodSystemGetProperty extends ValuePreservingMethod {
 /**
  * An access to a method named `getProperty` on class `java.lang.System`.
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/JDK.qll:262]:
```suggestion
 * A call to a method named `getProperty` on class `java.lang.System`.
```

File: java/ql/lib/semmle/code/java/Reflection.qll [hunk 5]
```diff
@@ -380,11 +395,14 @@ class ReflectiveAnnotationAccess extends ClassMethodAccess {
...
 
+/** DEPRECATED: Alias for `ReflectiveAnnotationCall`. */
+deprecated class ReflectiveAnnotationAccess = ReflectiveAnnotationCall;
+
 /**
  * A call to `Class.getField(..)` that accesses a field.
  */
-class ReflectiveFieldAccess extends ClassMethodAccess {
-  ReflectiveFieldAccess() {
+class ReflectiveFieldCall extends ClassMethodCall {
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Reflection.qll:404]: This class shouldn't be renamed. It's a call, yes, but it's a reflective access to a field, not a call to a field. The other subclasses of `ClassMethodCall` in this file should probably also not be renamed, as they're about accessing something via reflection and they're not calling whatever that is.

**Review by aschackmull — APPROVED** [review]
LGTM. I checked many MethodAccess renames, but not _all_ of them, as there were a lot.

**Review by aschackmull — DISMISSED** [review]

**Review by aschackmull — APPROVED** [review]

**Review by atorralba — APPROVED** [review]
Rubber-stamping @aschackmull's approval.

File: docs/codeql/ql-training/query-examples/java/data-flow-java-1.ql [hunk 6]
```diff
@@ -4,7 +4,7 @@ class StringConcat extends AddExpr {
   StringConcat() { getType() instanceof TypeString }
 }
 
-from MethodAccess ma
+from MethodCall ma
```
Comment (joefarebrother) [inline docs/codeql/ql-training/query-examples/java/data-flow-java-1.ql:7]: Should MethodCall variables named `ma` be renamed to `mc`?

Comment (smowton) [reply docs/codeql/ql-training/query-examples/java/data-flow-java-1.ql:7]: Ideally yes, but it's a harder transformation to script!

**Pull request merged**